src/node/txpackagetracker.cpp

```diff
@@ -0,0 +1,181 @@
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        auto count{orphan_request_tracker.CountInFlight(nodeid)};
+        return count;
+    }
+    size_t Count(NodeId nodeid) const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        auto count{orphan_request_tracker.Count(nodeid)};
+        return count;
+    }
+
+    std::vector<GenTxid> GetOrphanRequests(NodeId nodeid, std::chrono::microseconds current_time)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        std::vector<std::pair<NodeId, GenTxid>> expired;
+        auto tracker_requestable = orphan_request_tracker.GetRequestable(nodeid, current_time, &expired);
+        for (const auto& entry : expired) {
+            LogPrint(BCLog::TXPACKAGES, "\nTimeout of inflight %s %s from peer=%d\n", entry.second.IsWtxid() ? "ancpkginfo" : "orphan parent",
+                entry.second.GetHash().ToString(), entry.first);
+        }
+        std::vector<GenTxid> results;
+        for (const auto& gtxid : tracker_requestable) {
+            LogPrint(BCLog::TXPACKAGES, "\nResolving orphan %s, requesting by txids of parents from peer=%d\n", gtxid.GetHash().ToString(), nodeid);
+            const auto ptx = m_orphanage.GetTx(gtxid.GetHash());
+            if (!ptx) {
+                // We can't request ancpkginfo and we have no way of knowing what the missing
+                // parents are (it could also be that the orphan has already been resolved).
+                // Give up.
+                orphan_request_tracker.ForgetTxHash(gtxid.GetHash());
+                LogPrint(BCLog::TXPACKAGES, "\nForgetting orphan %s from peer=%d\n", gtxid.GetHash().ToString(), nodeid);
+                continue;
+            }
+            // Add the orphan's parents. Net processing will filter out what we already have.
+            // Deduplicate parent txids, so that we don't have to loop over
+            // the same parent txid more than once down below.
+            std::vector<uint256> unique_parents;
+            unique_parents.reserve(ptx->vin.size());
+            for (const auto& txin : ptx->vin) {
+                // We start with all parents, and then remove duplicates below.
+                unique_parents.push_back(txin.prevout.hash);
+            }
+            std::sort(unique_parents.begin(), unique_parents.end());
+            unique_parents.erase(std::unique(unique_parents.begin(), unique_parents.end()), unique_parents.end());
+            for (const auto& txid : unique_parents) {
+                results.emplace_back(GenTxid::Txid(txid));
+            }
+            // Mark the orphan as requested
+            orphan_request_tracker.RequestedTx(nodeid, gtxid.GetHash(), current_time + ORPHAN_ANCESTOR_GETDATA_INTERVAL);
+        }
+        if (!results.empty()) LogPrint(BCLog::TXPACKAGES, "\nRequesting %u items from peer=%d\n", results.size(), nodeid);
```

Comment:
This logging should be handled by the caller, since the caller sometimes isn't actually requesting the items!